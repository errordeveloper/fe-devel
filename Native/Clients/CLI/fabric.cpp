/*
 *  Copyright 2010-2011 Fabric Technologies Inc. All rights reserved.
 */

#include <Fabric/Clients/CLI/ClientFactory.h>
#include <Fabric/Core/Util/Log.h>
#include <Fabric/Core/Build.h>

#include <time.h>

namespace Fabric
{
  namespace CLI
  {
    extern "C" FABRIC_CLI_EXPORT void init( v8::Handle<v8::Object> target )
    {
      FABRIC_LOG( "%s version %s", Fabric::buildName, Fabric::buildFullVersion );
      struct tm const *lt = localtime( &Fabric::buildExpiry );
      char buf[1024];
      strftime( buf, 1024, "This build of Fabric will expire on %Y-%m-%d at %H:%M:%S", lt );
      FABRIC_LOG( "%s", buf );

      v8::HandleScope handleScope;
      target->Set( v8::String::New("createClient"), CreateClientV8Function() );
    }
  };
};
