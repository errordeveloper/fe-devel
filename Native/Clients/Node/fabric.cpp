/*
 *  Copyright 2010-2012 Fabric Engine Inc. All rights reserved.
 */

#include <Fabric/Clients/Node/ClientFactory.h>
#include <Fabric/Base/Util/Log.h>
#include <Fabric/Core/Build.h>

#include <llvm/Support/Threading.h>
#include <time.h>
#include <stdlib.h>

namespace Fabric
{
  namespace CLI
  {
    extern "C" FABRIC_CLI_EXPORT void init( v8::Handle<v8::Object> target )
    {
      FABRIC_LOG( "%s version %s", Fabric::buildName, Fabric::buildFullVersion );

      llvm::llvm_start_multithreaded();
      if ( !llvm::llvm_is_multithreaded() )
      {
        FABRIC_LOG( "LLVM not compiled with multithreading enabled; aborting" );
        abort();
      }

      v8::HandleScope handleScope;
      target->Set( v8::String::New("createClient"), CreateClientV8Function() );
    }
  }
}
