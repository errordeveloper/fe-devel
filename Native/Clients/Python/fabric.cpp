/*
 *  Copyright 2010-2012 Fabric Engine Inc. All rights reserved.
 */

#include <Fabric/Base/Util/SimpleString.h>
#include <Fabric/Base/Exception.h>
#include <Fabric/Clients/Python/ClientWrap.h>
#include <Fabric/Base/Util/Log.h>
#include <Fabric/Core/Build.h>

#include <map>
#include <time.h>
#include <llvm/Support/Threading.h>

//#define FABRIC_PYTHON_DEBUG

namespace Fabric
{
  namespace Python
  {
    extern "C" FABRIC_CLI_EXPORT void initialize()
    {
      FABRIC_LOG( "%s version %s", Fabric::buildName, Fabric::buildFullVersion );

      llvm::llvm_start_multithreaded();
      if ( !llvm::llvm_is_multithreaded() )
      {
        FABRIC_LOG( "LLVM not compiled with multithreading enabled; aborting" );
        abort();
      }
    }

    extern "C" FABRIC_CLI_EXPORT void createClient( void **ptr, char const *opts )
    {
      ClientWrap *client = new ClientWrap( opts );

#ifdef FABRIC_PYTHON_DEBUG
      FABRIC_LOG( "calling createClient: %lx", client);
#endif

      client->retain();
      *ptr = client;
    }
    
    extern "C" FABRIC_CLI_EXPORT void jsonExec( void *client_, char const *data, size_t length, const char **result )
    {
#ifdef FABRIC_PYTHON_DEBUG
      FABRIC_LOG( "calling jsonExec: %lx", client_ );
#endif

      ClientWrap *client( static_cast<ClientWrap *>( client_ ) );

#ifdef FABRIC_PYTHON_DEBUG
      FABRIC_LOG( "sending to jsonExec: %s", data );
#endif

      client->jsonExecAndAllocCStr( data, length, result );

#ifdef FABRIC_PYTHON_DEBUG
      FABRIC_LOG( "received from jsonExec: %s", *result );
#endif
    }

    extern "C" FABRIC_CLI_EXPORT void freeString( void *client_, void *string )
    {
#ifdef FABRIC_PYTHON_DEBUG
      FABRIC_LOG( "calling freeString: %lx (%lx)", client_, string );
#endif

      ClientWrap *client( static_cast<ClientWrap *>( client_ ) );
      client->freeJsonCStr( (const char *)string );
    }

    extern "C" FABRIC_CLI_EXPORT void setJSONNotifyCallback(
        void *client_,
        void (*callback)(const char *)
        )
    {
#ifdef FABRIC_PYTHON_DEBUG
      FABRIC_LOG( "calling setJSONNotifyCallback: %lx", client_ );
#endif

      ClientWrap *client( static_cast<ClientWrap *>( client_ ) );
      client->setJSONNotifyCallback( callback );
    }

    extern "C" FABRIC_CLI_EXPORT void freeClient( void *client_ )
    {
#ifdef FABRIC_PYTHON_DEBUG
      FABRIC_LOG( "calling close: %lx", client_ );
#endif

      ClientWrap *client( static_cast<ClientWrap *>( client_ ) );
      client->release();
    }

    extern "C" FABRIC_CLI_EXPORT void runScheduledCallbacks( void *client_ )
    {
#ifdef FABRIC_PYTHON_DEBUG
      FABRIC_LOG( "calling runScheduledCallbacks: %lx", client_ );
#endif

      ClientWrap *client( static_cast<ClientWrap *>( client_ ) );
      client->runScheduledCallbacks();
    }
  };
};
