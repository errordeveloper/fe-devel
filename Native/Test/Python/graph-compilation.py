#
#  Copyright 2010-2012 Fabric Engine Inc. All rights reserved.
#

import fabric

FABRIC = fabric.createClient()

def reportErrors( obj, expected ):
  numErrors = len( obj.getErrors() )
  if numErrors != expected:
    print( obj.getName() + " errors = " + str( numErrors ) +
      ", expected " + str( expected ) )

event = FABRIC.DependencyGraph.createEvent("bar")
event.fire()

reportErrors( event, 0 )

eventHandler = FABRIC.DependencyGraph.createEventHandler("baz")
event.appendEventHandler( eventHandler )
event.fire()

reportErrors( event, 0 )
reportErrors( eventHandler, 0 )

node = FABRIC.DependencyGraph.createNode("foo")
eventHandler.setScope( "self", node )
event.fire()

reportErrors( event, 0 )
reportErrors( eventHandler, 0 )
reportErrors( node, 0 )

op = FABRIC.DependencyGraph.createOperator("fooTwo")
binding = FABRIC.DependencyGraph.createBinding()
binding.setOperator( op )
binding.setParameterLayout( [ "self.foo" ] )
node.bindings.append( binding )

try:
  event.fire()
except Exception:
  print( "create and push operator: fire event successfully fails" )

reportErrors( event, 0 )
reportErrors( eventHandler, 0 )
reportErrors( node, 3 )

op.setEntryPoint("foo")
op.setSourceCode("operator foo( io Integer bar ) { bar = 7; }")

try:
  event.fire()
except Exception:
  print( "load operator bytecode: fire event successfully fails" )

reportErrors( event, 0 )
reportErrors( eventHandler, 0 )
reportErrors( node, 1 )

node.addMember( "foo", "Integer" )
event.fire()

reportErrors( event, 0 )
reportErrors( eventHandler, 0 )
reportErrors( node, 0 )

if node.getData('foo') != 7:
  print( "node.getData('foo') != 7" )

node.removeMember( "foo" )
try:
  event.fire()
except Exception:
  print( "remove node member: fire event successfully fails" )

reportErrors( event, 0 )
reportErrors( eventHandler, 0 )
reportErrors( node, 1 )

FABRIC.close()
