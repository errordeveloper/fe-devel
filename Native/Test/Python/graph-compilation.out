create and push operator: fire event successfully fails
load operator bytecode: fire event successfully fails
remove node member: fire event successfully fails
